client/app/widgets/QuickCreate/index.jsx

```diff
@@ -29,129 +25,170 @@ export type State = {
-    };
-  }
-
-  getCheckboxes = () => {
-    const { checkboxes } = this.state;
-    const checkboxesProp = [];
-    checkboxes.forEach((checkbox: Checkbox) => {
-      const checkboxProp = {
-        id: checkbox.id,
-        label: checkbox.label,
-        value: checkbox.value,
-        checked: checkbox.checked,
-      };
-      checkboxesProp.push(checkboxProp);
-    });
-    return checkboxesProp;
-  };
+export const labelExists = (checkboxes: Checkbox[], compareLabel: string) => checkboxes.filter(
+  (checkbox: Checkbox) => checkbox.label.toLowerCase() === compareLabel.toLowerCase(),
+).length;
 
-  labelExists = (label: string) => {
-    const { checkboxes } = this.state;
-    return checkboxes.filter(
-      (checkbox: Checkbox) => checkbox.label.toLowerCase() === label.toLowerCase(),
-    ).length;
-  };
+export const addToCheckboxes = (
+  { name, id, slug }: NewCheckbox,
+  checkboxes: Checkbox[],
+) => {
+  const newCheckboxes = checkboxes.slice(0);
+  newCheckboxes.push({
+    id: slug,
+    label: name,
+    value: id,
+    checked: true,
+  });
+  return sortAlpha(newCheckboxes);
+};
 
-  addToCheckboxes = (data: { name: string, id: string, slug: string }) => {
-    const { checkboxes } = this.state;
-    const { name, id, slug } = data;
-    const newCheckboxes = checkboxes.slice(0);
-    newCheckboxes.push({
-      id: slug,
-      label: name,
-      value: id,
-      checked: true,
-    });
-    return sortAlpha(newCheckboxes);
-  };
+// reducer types
+type OnChangePayload = {
+  label: string,
+  checkboxes: Checkbox[],
+  body: any,
+};
+type CheckboxData = {
+  name: string,
+  id: string,
+  slug: string,
+  success?: boolean,
+};
+type OnSubmitPayload = CheckboxData;
+type OnChangeAction = { type: 'ON_CHANGE', payload: OnChangePayload };
+type OnSubmitAction = { type: 'ON_SUBMIT', payload: OnSubmitPayload };
+type Action = OnChangeAction | OnSubmitAction;
+// reducer constants
+export const ON_CHANGE = 'ON_CHANGE';
+export const ON_SUBMIT = 'ON_SUBMIT';
 
-  onSubmit = (response: any) => {
-    const { data } = response;
-    if (data && data.success) {
-      this.setState({
+const quickCreateReducer = (state: State, action: Action) => {
+  switch (action.type) {
+    case ON_CHANGE: {
+      return {
+        ...state,
+        checkboxes: action.payload.checkboxes,
+        label: action.payload.label,
+        open: true,
+        modalKey: Utils.randomString(),
+        body: action.payload.body,
+      };
+    }
+    case ON_SUBMIT: {
+      return {
+        ...state,
         open: false,
         accordionOpen: true,
-        modalKey: Utils.randomString(),
         tagKey: Utils.randomString(),
-        checkboxes: this.addToCheckboxes(data),
-      });
+        modalKey: Utils.randomString(),
+        checkboxes: addToCheckboxes(action.payload, state.checkboxes),
+      };
     }
-  };
+    default:
+      throw new Error(`Unhandled action type: ${action.type}`);
+  }
+};
```

Comment:
I struggled a bit with this to be honest. What I observed when I had consecutive state updates for onSubmit or onChange was: old checkbox state ->  new tagKey state -> new checkbox state
This was a problem because the tagKey is forcing a brand new input tag to be rendered, and in our case it was being forced to rerender with the now stale checkbox state. 

The only way I got around this solely with state hooks was using a boolean state flag that was used as a dependency in an effect. If I recall correctly, state updates ran at that point reflected the new checkbox state. This was kind of smelly so I looked around for other options. 

I had reviewed https://overreacted.io/a-complete-guide-to-useeffect/ and https://kentcdodds.com/blog/should-i-usestate-or-usereducer then came to the conclusion that this was a decent use case for useReducer for a few reasons. 

1.  tagKey and checkboxes are coupled in that they must update together else child components don't reflect the current state. 
2. Functionally, this component has two primary responsibilities (onChange and onSubmit) and when those responsibilities are invoked they have a set of specific required state updates. The specific state updates and necessary steps are nicely encapsulated in the reducer.

Reply:
For sure, I liked the abstraction you've used for the same reasons.

Oh interesting, I wonder if there's a way to resolve that with `useState`, but I see why you grouped those states together. Will post an update here in case I find anything useful there 👍

Reply:
I appreciate you taking a deeper look at this. Yeah at some point I also realized ah! `tagKey` is referring to the key value on the input of type tag that forces it to remount so that it is initialized with checkboxes prop. 
Although I did not make the connection to wrap the update of the tag with an effect having dependency on checkboxes. 

I think sometimes I am hesitant to use objects/arrays as the dependencies because it's usually a specific value I'm wanting to run the effect against. If that specific value that I am interested in changes does that also mean the object/array reference has changed? That's something I can't answer confidently. 

Based on your example, and the effect of it, it seems like it is the case that there would be a new object/array reference.

I'm less inclined to change this for a few reasons. One as you mentioned, `tagKey` updated onChange. Two, tagKey and checkboxes need to update together, so if we colocate that update it could be easier to read and less likely to be pulled apart if the requirements of the component change. Three the reducer actions themselves encapsulate all the work that needs to be done onChange or onSubmit.